src/codecs/ready_frames.rs

```diff
@@ -0,0 +1,113 @@
+use std::pin::Pin;
+
+use futures::{
+    task::{Context, Poll},
+    {Stream, StreamExt},
+};
+
+const DEFAULT_CAPACITY: usize = 1024;
+
+/// A stream combinator aimed at improving the performance of decoder streams under load.
+///
+/// This is similar in spirit to `StreamExt::ready_chunks`, but built specifically for the
+/// particular result tuple returned by decoding streams. The more general `FoldReady` is left as
+/// an exercise to the reader.
+pub struct ReadyFrames<T, U, E> {
+    inner: T,
+    enqueued: Vec<U>,
+    enqueued_size: usize,
+    error_slot: Option<E>,
+    enqueued_limit: usize,
+}
+
+impl<T, U, E> ReadyFrames<T, U, E>
+where
+    T: Stream<Item = Result<(U, usize), E>> + Unpin,
+    U: Unpin,
+    E: Unpin,
+{
+    /// Create a new `ReadyChunks` by wrapping a decoder stream, most commonly a `FramedRead`.
+    pub fn new(inner: T) -> Self {
+        Self::with_capacity(inner, DEFAULT_CAPACITY)
+    }
+
+    /// Create a new `ReadyChunks` with a specified capacity by wrapping a decoder stream, most
+    /// commonly a `FramedRead`.
+    ///
+    /// The specified capacity is a soft limit, and chunks may be returned that exceed that size.
```

Comment:
```suggestion
    /// The specified capacity is a soft limit, and chunks may be returned that exceed that length.
```

This is only a minor suggestion, but I had to triple read through the code to convince myself that `enqueued_limit` always refers to "length" rather than "byte size" and is used accordingly